feat: Split Github Actions & Cache Runs

H4ad opened the pull request:
Closes #32

With these changes, I was able to run this workflow: 

![image](https://github.com/RafaelGSS/nodejs-bench-operations/assets/12551007/fc02fd4c-7d6a-4a15-bfa4-88d4dfd0eeee)
> https://github.com/h4ad-forks/nodejs-bench-operations/actions/runs/5581804553

The workflows created were:
- `watch_bench.yml`: Which runs benchmarks on modified files
- `bench.yml`: This allows running a specific benchmark.

What looks promising:
- it's very fast to run since only what is being modified is running
- it will be quick to update the report when a new benchmark comes in, instead of running everything, it runs only for those that entered
- I added an option to run manually, which allows you to update the reports whenever you want

Current Tasks:
- [x] split benchmarks
- [x] cache the result of those benchmarks
- [x] create script that generates the final report, joining all partial reports
- [x] create script that generates the major report (eg: RESULTS-v20.md)
- [x] add machine specifications to every benchmark report
- [x] add support for skipping benchmarks, for when you want to do something that's just lint

RafaelGSS reviewed (commented):
> there is a limit of 256 jobs that can be run, and the number of benchmarks that supports generating the report at a time is 256 / Number of Node Versions, which currently gives: 19 benchmarks

As I've said previously, we can't parallel benchmarks. We can use only the benchmark the diff, but it must be sequential

H4ad commented on .github/workflows/bench.yml:
```diff
@@ -0,0 +1,100 @@
+    strategy:
+      matrix:
+        node-version: [16.0.0, 16.18.1, 16.19.0, 16.20.0, 18.0.0, 18.14.2, 18.15.0, 18.16.0, 19.0.0, 19.8.0, 19.8.1, 19.9.0, 20.0.0]
+      max-parallel: 4
```
Since the machine will only take one job at time, this value doesn't matter.

But I will set this value as `1`, just to ensure to run one benchmark at time.

H4ad commented:
One question about these details that are concatenated on every report:

```
# Node.js Benchmark Operations

* __Machine:__ linux x64 | 4 vCPUs | 7.3GB Mem
* __Node:__ `v16.20.0`
* __Run:__ Tue Jul 11 2023 22:16:22 GMT+0000 (Coordinated Universal Time)
```

I was thinking in add them on every benchmark report like this:

<details>
<summary>Environment</summary>

* __Machine:__ linux x64 | 4 vCPUs | 7.3GB Mem
* __Node:__ `v16.20.0`
* __Run:__ Tue Jul 11 2023 22:16:22 GMT+0000 (Coordinated Universal Time)
</details>

Also, to do that, the easiest way is to modify every benchmark file to emit that information after finishing the benchmark, should I do that?

RafaelGSS commented:
> Also, to do that, the easiest way is to modify every benchmark file to emit that information after finishing the benchmark, should I do that?

I don't think so. You can add a step earlier in the workflow that will generate this output

H4ad commented:
Added the machine info on every benchmark, I also added the main report based on the latest node version.

About the Skip CI, I found this: https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs
So, I don't think I need to support something directly in the action.

Now, this PR is good to be reviewed.

RafaelGSS commented on .github/workflows/bench.yml:
```diff
@@ -0,0 +1,102 @@
+          result=$(cat ./bench-result.md)
+          echo "$result" >> $GITHUB_STEP_SUMMARY
+          export RESULT_IN_BASE64=$(cat ./bench-result.md | base64 -w 0)
+          echo "BENCH_RESULT=$RESULT_IN_BASE64" >> "$GITHUB_ENV"
```
I think it's better to run it in your fork first. I believe BASE64 will skip \n\t\r

RafaelGSS reviewed (commented):
Could you run all the workflow in your fork first? I'm a bit skeptical this will work in the way you think it gonna work.

H4ad commented:
Run: https://github.com/h4ad-forks/nodejs-bench-operations/actions/runs/5872774390
The commit with the changes: https://github.com/h4ad-forks/nodejs-bench-operations/commit/5736a5777f9a4215b2cc48872ab7f0dcf02af581
